src/buffer/out/textBuffer.cpp

```diff
@@ -166,6 +166,29 @@ ROW& TextBuffer::_getRowByOffsetDirect(size_t offset)
     return *reinterpret_cast<ROW*>(row);
 }
 
+// This function is "direct" because it trusts the caller to know that this row
+// could be the scratchpad row.
+size_t TextBuffer::_estimateOffsetOfLastCommittedRowDirect() const
+{
+    if (_commitWatermark == _buffer.get())
+    {
+        // Estimation failed. Act as though we have at least one non-scratchpad row committed.
+        // The caller will be passing this offset back to GetRowByOffset or similar later, which
+        // will commit it.
+        return 1;
+    }
+    const auto lastRowOffset = static_cast<ptrdiff_t>((_commitWatermark - 1) - _buffer.get()) / _bufferRowStride;
+    return gsl::narrow_cast<size_t>(lastRowOffset);
+}
+
+// Returns the "user-visible" index of the last committed row, which can be used
+// to short-circuit some algorithms that try to scan the entire buffer..
+size_t TextBuffer::_estimateOffsetOfLastCommittedRow() const
+{
+    // Take into account the scratchpad row.
+    return ((_estimateOffsetOfLastCommittedRowDirect() + _height - 1) % _height);

```

Comment:
I could have probably simplified this logic because `_estimateOffsetDirect` will never return 0. This was absolutely intended to handle it returning 0.